pkg/distributor/distributor.go

```diff
@@ -859,6 +861,19 @@ func (d *Distributor) prepareSeriesKeys(ctx context.Context, req *cortexpb.Write
 	// check each sample and discard if outside limits.
 	skipLabelNameValidation := d.cfg.SkipLabelNameValidation || req.GetSkipLabelNameValidation()
 	for _, ts := range req.Timeseries {
+		removeReplicaTs := removeReplica
+		if limits.AcceptHASamples && limits.AcceptMixedHASamples {
+			validCode, cluster := d.isValidHAPair(ts.Labels, ctx, userID, limits)
+			if validCode == 1 {
+				d.dedupedSamples.WithLabelValues(userID, cluster).Add(float64(len(ts.Samples) + len(ts.Histograms)))
+				continue
+			} else if validCode == 2 {
+				removeReplicaTs = false
+				d.nonHASamples.WithLabelValues(userID).Add(float64(len(ts.Samples) + len(ts.Histograms)))
+			} else if validCode == 0 {
+				removeReplicaTs = true
+			}
+		}
```

Comment:
```suggestion
		if limits.AcceptHASamples && limits.AcceptMixedHASamples {
			cluster, replica := findHALabels(limits.HAReplicaLabel, limits.HAClusterLabel, ts.Labels)
			if cluster != "" && replica != "" {
				_, err := d.checkSample(ctx, userID, cluster, replica, limits)
				if err != nil {
					// discard sample
					d.dedupedSamples.WithLabelValues(userID, cluster).Add(float64(len(ts.Samples) + len(ts.Histograms)))
					continue
				}
				removeReplica = true // valid HA sample
			}
			removeReplica = false
			d.nonHASamples.WithLabelValues(userID).Add(float64(len(ts.Samples) + len(ts.Histograms)))
		}
```

How about deleting `isValidHAPair` and `removeReplicaTs`? 
WDYT

Reply:
Yes, this is more Go style with the err returned. Will do!